# (PART) PHYLOSEQ OBJECTS {-}

# Creating phyloseq Objects

In this section, we demonstrate how to create phyloseq objects to organize and structure microbiome data for analysis. These objects serve as the foundation for downstream analyses and visualization.

## From mothur output
```{r mothurphyseqobj, message=FALSE, warning=FALSE}
library(tidyverse)
library(phyloseq)
library(microbiome)
library(ape)

metadata <- read_csv("data/mothur/mothur_tidy_metadata.csv",show_col_types = FALSE) %>% 
  tibble::column_to_rownames("group") %>% 
  sample_data(metadata)

otutable <- read_csv("data/mothur/mothur_tidy_otutable.csv",show_col_types = FALSE) %>% 
  pivot_wider(id_cols = group, names_from = OTU, values_from = count) %>% 
  tibble::column_to_rownames("group") %>% 
  otu_table(otutable, taxa_are_rows = FALSE)

taxonomy <- read_csv("data/mothur/mothur_tidy_taxonomy.csv", show_col_types = FALSE) %>%
  tibble::column_to_rownames("OTU") %>%
  as.matrix() %>% 
  tax_table(taxonomy)

mt_ps_raw_basic <- merge_phyloseq(metadata, otutable, taxonomy)

library(ape)
mt_ps_tree = rtree(ntaxa(mt_ps_raw_basic), rooted=TRUE, tip.label=taxa_names(mt_ps_raw_basic))

mt_ps_raw <- phyloseq::merge_phyloseq(mt_ps_raw_basic, mt_ps_tree)

mt_ps_rel <- phyloseq::transform_sample_counts(mt_ps_raw, function(x){x / sum(x)})

mt_ps_df_raw <- psmelt(mt_ps_raw)

mt_ps_df_rel <- psmelt(mt_ps_raw)

save(mt_ps_tree, mt_ps_raw, mt_ps_rel, mt_ps_df_raw, mt_ps_df_rel, file = "data/mothur_phyloseq_objects.rda")

```


### Review the Mothur phyloseq object

```{r mothurpsraw}
mt_ps_raw
```

## From QIIME2 output
```{r qiime2physeqobj, message=FALSE, warning=FALSE}
library(tidyverse)
library(phyloseq)
library(microbiome)
library(ape)

metadata <- read_csv("data/qiime2/qiime2_tidy_metadata.csv",show_col_types = FALSE) %>% 
  tibble::column_to_rownames("sample_id") %>% 
  sample_data(metadata)

otutable <- read_csv("data/qiime2/qiime2_tidy_otutable.csv",show_col_types = FALSE) %>% 
  pivot_wider(id_cols = sample_id, names_from = feature, values_from = count) %>% 
  tibble::column_to_rownames("sample_id") %>% 
  otu_table(otutable, taxa_are_rows = FALSE)

taxonomy <- read_csv("data/qiime2/qiime2_tidy_taxonomy.csv", show_col_types = FALSE) %>%
  tibble::column_to_rownames("feature") %>%
  as.matrix() %>% 
  tax_table(taxonomy)

q2_ps_raw_basic <- merge_phyloseq(metadata, otutable, taxonomy)

library(ape)
q2_ps_tree = rtree(ntaxa(q2_ps_raw_basic), rooted=TRUE, tip.label=taxa_names(q2_ps_raw_basic))

q2_ps_raw <- phyloseq::merge_phyloseq(q2_ps_raw_basic, q2_ps_tree)

q2_ps_rel <- phyloseq::transform_sample_counts(q2_ps_raw, function(x){x / sum(x)})

q2_ps_df_raw <- psmelt(q2_ps_raw)

q2_ps_df_rel <- psmelt(q2_ps_raw)

save(q2_ps_tree, q2_ps_raw, q2_ps_rel, q2_ps_df_raw, q2_ps_df_rel, file = "data/qiime2_phyloseq_objects.rda")
```


## Review the QIIME2 phyloseq object

```{r qiime2psraw}
q2_ps_raw
```



# (PART) EXTERNAL DATASETS {-}

# Phyloseq Objects Demo Datasets
In this section, we utilize publicly available real-world datasets to demonstrate the functionality and analysis workflows of phyloseq objects. These datasets serve as valuable resources for researchers seeking to reproduce and validate their analyses using similar microbiome data. By leveraging these demo datasets, we aim to enhance reproducibility and accessibility in microbiome research and analysis.

```{r, include=FALSE}
knitr::opts_chunk$set(
  echo  =TRUE,
  message  =FALSE,
  warning  =FALSE,
  cache  =FALSE,
  comment  =NA)

library(tidyverse)
```


## The GlobalPatterns dataset
The GlobalPatterns dataset, sourced from the Earth Microbiome Project (EMP), serves as a comprehensive repository for studying microbial communities worldwide. Here's a breakdown of its key attributes:

- **Source:** GlobalPatterns originates from the Earth Microbiome Project (EMP), collecting samples worldwide.

- **Composition:** It comprises high-throughput sequencing data, revealing the taxonomic composition of microbial communities.

- **Scope:** Samples represent diverse global ecosystems, offering a comprehensive view of microbial biodiversity.

- **Format:** Presented as a phyloseq object in R, it integrates sample metadata and taxonomic abundance for analysis.

- **Utility:** Researchers utilize it for community profiling, abundance testing, and ecological modeling, enhancing understanding of global microbial diversity and function.

```{r}
library(phyloseq) # for GlobalPatterns dataset
data("GlobalPatterns")

ps_GlobalPatterns <-GlobalPatterns
df_GlobalPatterns <-GlobalPatterns %>% 
  phyloseq::psmelt() %>% 
  tibble::rownames_to_column("sample_id") %>% 
  rename_all(tolower)
```

**GlobalPatterns phyloseq-class**

```{r}
ps_GlobalPatterns
```

**Columns in GlobalPatterns  dataset**

```{r}
colnames(df_GlobalPatterns)
cat("\n")

```

## The dietswap dataset
The dietswap dataset, available through the microbiome R package, offers valuable insights into the effects of dietary interventions on microbial communities. Here's an overview of its attributes:

- **Source:** The dietswap dataset is derived from research examining the impact of dietary changes on the human gut microbiome.

- **Composition:** It comprises high-throughput sequencing data, providing insights into the taxonomic composition and dynamics of microbial communities in response to dietary alterations.

- **Scope:** Samples are obtained from human participants undergoing dietary interventions, enabling researchers to explore how different diets influence microbial diversity and function within the gut microbiome.

- **Format:** Presented in a format suitable for microbiome analysis, the dataset includes sample metadata and taxonomic abundance data, facilitating comprehensive analyses of microbial community dynamics.

- **Utility:** Researchers utilize the dietswap dataset to investigate the effects of dietary interventions on gut microbiome composition, contributing to a better understanding of the intricate interactions between diet, host physiology, and microbial ecology.

```{r}
library(microbiome) # for dietswap dataset
data("dietswap")

dietswap
```

> Note: dietswap is missing the phylo_tree slot, we can construct it and add it like so:

```{r message=FALSE, warning=FALSE}
library(microbiome)
data('dietswap')
ps_raw_basic <- dietswap

library(ape)
ps_tree = rtree(ntaxa(ps_raw_basic), rooted=TRUE, tip.label=taxa_names(ps_raw_basic))
ps_dietswap <- phyloseq::merge_phyloseq(ps_raw_basic, ps_tree)

df_dietswap <-ps_dietswap %>% 
  phyloseq::psmelt() %>% 
  tibble::rownames_to_column("sample_id") %>% 
  dplyr::select(-9) %>% 
  rename_all(tolower)

```


**Dietswap phyloseq-class**

```{r}
ps_dietswap
```

**Columns in Dietswap  dataset**

```{r}
colnames(df_dietswap)

```

## Caporaso dataset

The Caporaso dataset provides valuable insights into microbial marker genes and their associations with various biological factors.

- **Source:** The Caporaso dataset is included in the `microbiomeMarker` R package and is derived from research led by Dr. J. Gregory Caporaso. It aims to provide insights into microbial marker genes and their associations with various biological factors.

- **Composition:** The dataset comprises high-throughput sequencing data, focusing on microbial marker genes from diverse biological samples. It offers valuable information regarding the taxonomic composition and functional potential of microbial communities.

- **Scope:** Samples are collected from a range of environments, including but not limited to human microbiomes, environmental samples, and animal microbiomes. This diversity allows researchers to explore microbial diversity across different ecosystems and conditions.

- **Format:** The dataset is structured to facilitate microbiome marker analysis, with sample metadata and taxonomic abundance data included. This format enables researchers to conduct comprehensive analyses of microbial marker genes and their associations with environmental or biological factors.

- **Utility:** Researchers utilize the Caporaso dataset to investigate microbial marker genes' roles in various ecosystems, such as host-microbiome interactions, environmental responses, and disease states. The dataset contributes to a better understanding of microbial ecology and its implications for human health and environmental management.

```{r message=FALSE, warning=FALSE}
library(microbiomeMarker) # for caporaso dataset
data("caporaso")

ps_caporaso <-caporaso
df_caporaso <-caporaso %>% 
  phyloseq::psmelt() %>% 
  tibble::rownames_to_column("sample_id") %>% 
  rename_all(tolower)

```


**Caporaso phyloseq-class**

```{r}
ps_caporaso
```

**Columns in Caporaso  dataset**

```{r}
colnames(df_caporaso)

```


## Kostic_CRC dataset

The Kostic_CRC dataset provides valuable insights into the gut microbiome of individuals diagnosed with colorectal cancer (CRC).

- **Source:** The Kostic_CRC dataset is included in the `microbiomeMarker` R package. It is derived from research conducted by the Kostic Lab and focuses on investigating the gut microbiome in colorectal cancer (CRC) patients.

- **Composition:** This dataset comprises high-throughput sequencing data, specifically targeting the gut microbiome of individuals with colorectal cancer. It provides insights into the taxonomic composition and potential functional characteristics of microbial communities associated with CRC.

- **Scope:** Samples are collected from individuals diagnosed with colorectal cancer, allowing researchers to explore the microbial diversity and potential biomarkers associated with CRC development and progression.

- **Format:** The dataset is structured to facilitate microbiome marker analysis, including sample metadata and taxonomic abundance data. This format enables researchers to conduct comprehensive analyses of microbial community dynamics in colorectal cancer.

- **Utility:** Researchers utilize the Kostic_CRC dataset to investigate the role of the gut microbiome in colorectal cancer pathogenesis, prognosis, and treatment response. The dataset contributes to a deeper understanding of the complex interplay between the gut microbiome and colorectal cancer biology, potentially leading to novel diagnostic or therapeutic strategies.


```{r message=FALSE, warning=FALSE}
library(microbiomeMarker) # for kostic_crc dataset
data("kostic_crc")

ps_raw_basic <- kostic_crc

library(ape)
ps_tree = rtree(ntaxa(ps_raw_basic), rooted=TRUE, tip.label=taxa_names(ps_raw_basic))
ps_kostic_crc <- phyloseq::merge_phyloseq(ps_raw_basic, ps_tree)

df_kostic_crc <-kostic_crc %>% 
  phyloseq::psmelt() %>% 
  tibble::rownames_to_column("sample_id") %>% 
  rename_all(tolower)

```


**Kostic_crc phyloseq-class**

```{r}
ps_kostic_crc
```

**Columns in Kostic_crc  dataset**

```{r}
colnames(df_kostic_crc)

```

## Save objects for transformation and exploration {#save_ps_objects}

```{r}
save(df_GlobalPatterns, 
     df_dietswap,  
     df_caporaso,
     df_kostic_crc,     
     file = "data/dataframe_objects.rda")

save(ps_GlobalPatterns, 
     ps_dietswap,
     ps_caporaso,
     ps_kostic_crc,
     file = "data/phyloseq_objects.rda")

```

## Confirm saved objects
```{r}
load("data/phyloseq_objects.rda", verbose = TRUE)
```


# (PART) PHYLOSEQ FUNCTIONS {-}

# Reviewing Phyloseq Objects
In this section, we demonstrate basic functions for reviewing and exploring the contents of phyloseq objects. By familiarizing ourselves with these functions, we gain insights into the structure, composition, and metadata associated with phyloseq objects, facilitating effective analysis and interpretation of microbiome data.

```{r message=FALSE, warning=FALSE}
library(phyloseq)
library(ape)

ps <- ps_dietswap
ps
```

## Taxonomy rank names
```{r}
rank_names(ps)

```

## Number of taxa
```{r}
ntaxa(ps)
```

## Split by taxon name, e.g. Firmicutes
```{r message=FALSE, warning=FALSE}
library(metagMisc) # foe phyloseq_sep_tax()

taxa_stats <- phyloseq_sep_tax(ps, TaxRank = "Phylum", drop_NA = FALSE)
cat("\n")
taxa_stats$Firmicutes
cat("\n")
taxa_stats$Bacteroidetes
cat("\n")
taxa_stats$Actinobacteria

```


## Number of samples
```{r}
nsamples(ps)
```

## Sample names
```{r}
head(sample_names(ps), 20)
```


## Sample variables
```{r}
sample_variables(ps)
```

## Abundance table
```{r}
otu_table(ps)[1:5, 1:5]
```

## Taxonomy table
```{r}
tax_table(ps)[1:5, ]
```

## Phylogenetic tree if available
```{r, warning=FALSE, include=FALSE}
# phy_tree(ps)
```

## Taxa names
```{r}
taxa_names(ps)[1:10]
```

## Summarize Phyloseq Object
```{r message=FALSE, warning=FALSE}

microbiome::summarize_phyloseq(ps)

```

## Sort Samples in ascending and descending
```{r}
cat("In ascending order\n")
head(sort(phyloseq::sample_sums(ps), decreasing = F))
cat("\nIn descending order\n")
head(sort(phyloseq::sample_sums(ps), decreasing = T))

```

## Drop Samples Below a Threshold
```{r message=FALSE, warning=FALSE}
ps1perc0 <- phyloseq::subset_samples(ps, phyloseq::sample_sums(ps) > 1)
ps1perc0
```

## Drop Taxa Below a Threshold
```{r message=FALSE, warning=FALSE}
pstaxa01 <- phyloseq::prune_taxa(phyloseq::taxa_sums(ps) > 1, ps)
pstaxa01 
```


# (PART) PHYLOSEQ EXTRA OBJECTS {-}

# Create psExtra Objects

After the creation of phyloseq objects, we proceed to create psExtra objects using functionalities provided by the **microViz** package. The psExtra objects extend the capabilities of phyloseq objects by incorporating additional metadata, experimental conditions, or statistical results, enhancing the depth and breadth of analyses performed using microbiome data.

## Prerequisite

Before creating psExtra objects, it's essential to have pre-existing phyloseq objects containing microbiome data and associated metadata. These phyloseq objects serve as the input data for creating psExtra objects. For example, pre-existing phyloseq objects such as ps_dietswap, ps_GlobalPatterns, ps_caporaso, and ps_kostic_crc prepared in the previous sections should be available.

## Load pre-existing phyloseq objects
```{r}
load("data/phyloseq_objects.rda", verbose = TRUE)
```

## Aggregate taxonomic data
```{r message=FALSE, warning=FALSE}
# Load required packages
library(phyloseq) # For handling microbiome data
library(microViz) # For advanced microbiome visualization

# Assuming ps_dietswap is already loaded or available

# Aggregate taxonomic data
psextra_clr_dietswap <- tax_agg(
  ps = ps_dietswap,
  rank = "Genus") %>% 
  tax_transform(trans = "clr")

# Display a subset of the transformed data
cat("\nCentered log Ratio (CLR) Transformation\n")
psextra_clr_dietswap %>% 
  otu_get() %>%
  .[1:4, 1:3]

# Aggregate taxonomic data
psextra_id_dietswap <- tax_agg(
  ps = ps_dietswap,
  rank = "Genus") %>% 
  tax_transform(trans = "identity")

# Display a subset of the transformed data
cat("\nIdentity Transformation\n")
psextra_id_dietswap %>% 
  otu_get() %>%
  .[1:4, 1:3]

# Aggregate taxonomic data
psextra_log10p_dietswap <- tax_agg(
  ps = ps_dietswap,
  rank = "Genus") %>% 
  tax_transform(trans = "log10p")

# Display a subset of the transformed data
cat("\nLog10p Transformation\n")
psextra_log10p_dietswap %>% 
  otu_get() %>%
  .[1:4, 1:3]

```


## Confirm the psExtra class
```{r class}
summary(psextra_clr_dietswap)
summary(psextra_id_dietswap)
summary(psextra_log10p_dietswap)
```

## Save psExtra to R object
```{r}
save(psextra_clr_dietswap, psextra_id_dietswap, psextra_log10p_dietswap, file = "data/phyloseq_extra_objects.rda")

```


