<!DOCTYPE html>
<html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>9 Data Tranformation from Phyloseq Objects | IMAP: Integrated Microbiome Analysis Pipelines</title>
<meta name="author" content="Teresia Mrema-Buza">
<meta name="description" content="The process of data transformation in microbiome analysis involves converting raw or relative abundance values into matrices that are suitable for further analysis. Data transformation focuses on...">
<meta name="generator" content="bookdown 0.37 with bs4_book()">
<meta property="og:title" content="9 Data Tranformation from Phyloseq Objects | IMAP: Integrated Microbiome Analysis Pipelines">
<meta property="og:type" content="book">
<meta property="og:url" content="https://tmbuza.github.io/imap-data-preparation/data-tarnsformation.html">
<meta property="og:image" content="https://tmbuza.github.io/imap-data-preparation/images/planning.png">
<meta property="og:description" content="The process of data transformation in microbiome analysis involves converting raw or relative abundance values into matrices that are suitable for further analysis. Data transformation focuses on...">
<meta name="twitter:card" content="summary">
<meta name="twitter:title" content="9 Data Tranformation from Phyloseq Objects | IMAP: Integrated Microbiome Analysis Pipelines">
<meta name="twitter:description" content="The process of data transformation in microbiome analysis involves converting raw or relative abundance values into matrices that are suitable for further analysis. Data transformation focuses on...">
<meta name="twitter:image" content="https://tmbuza.github.io/imap-data-preparation/images/planning.png">
<!-- JS --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://kit.fontawesome.com/6ecbd6c532.js" crossorigin="anonymous"></script><script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="libs/bootstrap-4.6.0/bootstrap.min.css" rel="stylesheet">
<script src="libs/bootstrap-4.6.0/bootstrap.bundle.min.js"></script><link href="libs/Sen-0.4.8/font.css" rel="stylesheet">
<link href="https://fonts.googleapis.com/css2?family=DM%20Mono&amp;display=swap" rel="stylesheet">
<link href="libs/Bitter-0.4.8/font.css" rel="stylesheet">
<script src="libs/bs3compat-0.6.1/transition.js"></script><script src="libs/bs3compat-0.6.1/tabs.js"></script><script src="libs/bs3compat-0.6.1/bs3compat.js"></script><link href="libs/bs4_book-1.0.0/bs4_book.css" rel="stylesheet">
<script src="libs/bs4_book-1.0.0/bs4_book.js"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- CSS --><style type="text/css">
    
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  </style>
<link rel="stylesheet" href="style.css">
</head>
<body data-spy="scroll" data-target="#toc">

<div class="container-fluid">
<div class="row">
  <header class="col-sm-12 col-lg-3 sidebar sidebar-book"><a class="sr-only sr-only-focusable" href="#content">Skip to main content</a>

    <div class="d-flex align-items-start justify-content-between">
      <h1>
        <a href="index.html" title="End-to-End Practical User Guides Using Integrated Approaches"><big>IMAP:</big> Integrated Microbiome Analysis Pipelines</a>:
        <small class="text-muted">End-to-End Practical User Guides Using Integrated Approaches</small>
      </h1>
      <button class="btn btn-outline-primary d-lg-none ml-2 mt-1" type="button" data-toggle="collapse" data-target="#main-nav" aria-expanded="true" aria-controls="main-nav"><i class="fas fa-bars"></i><span class="sr-only">Show table of contents</span></button>
    </div>

    <div id="main-nav" class="collapse-lg">
      <form role="search">
        <input id="search" class="form-control" type="search" placeholder="Search" aria-label="Search">
</form>

      <nav aria-label="Table of contents"><h2>Table of contents</h2>
        <ul class="book-toc list-unstyled">
<li><a class="" href="index.html">IMAP-Part 07: Microbiome Data Preparation Guide</a></li>
<li class="book-part">GETTING STARTED</li>
<li><a class="" href="preface.html">Preparing for Downstream Analysis</a></li>
<li><a class="" href="data-tidying.html"><span class="header-section-number">1</span> Data Tidying</a></li>
<li class="book-part">MOTHUR OUTPUT</li>
<li><a class="" href="prepare-mothur-output.html"><span class="header-section-number">2</span> Prepare Mothur Output</a></li>
<li class="book-part">QIIME2 OUTPUT</li>
<li><a class="" href="prepare-qiime2-output.html"><span class="header-section-number">3</span> Prepare Qiime2 Output</a></li>
<li class="book-part">PHYLOSEQ OBJECTS</li>
<li><a class="" href="creating-phyloseq-objects.html"><span class="header-section-number">4</span> Creating phyloseq Objects</a></li>
<li class="book-part">EXTERNAL DATASETS</li>
<li><a class="" href="phyloseq-objects-demo-datasets.html"><span class="header-section-number">5</span> Phyloseq Objects Demo Datasets</a></li>
<li class="book-part">PHYLOSEQ FUNCTIONS</li>
<li><a class="" href="reviewing-phyloseq-objects.html"><span class="header-section-number">6</span> Reviewing Phyloseq Objects</a></li>
<li class="book-part">PHYLOSEQ EXTRA OBJECTS</li>
<li><a class="" href="create-psextra-objects.html"><span class="header-section-number">7</span> Create psExtra Objects</a></li>
<li class="book-part">CLEANING PHYLOSEQ OBJECT</li>
<li><a class="" href="pipeline-for-cleaning-phyloseq-objects.html"><span class="header-section-number">8</span> Pipeline for Cleaning Phyloseq Objects</a></li>
<li class="book-part">DATA TRANSFORMATION</li>
<li><a class="active" href="data-tarnsformation.html"><span class="header-section-number">9</span> Data Tranformation from Phyloseq Objects</a></li>
<li class="book-part">DISTANCES MATRICES</li>
<li><a class="" href="distance-calculation-in-microbiome-data-analysis.html"><span class="header-section-number">10</span> Distance Calculation in Microbiome Data Analysis</a></li>
<li><a class="" href="utilizing-transformed-data.html"><span class="header-section-number">11</span> Whatâ€™s Next: Utilizing Transformed Data</a></li>
<li class="book-part">DIMENSIONALITY REDUCTION</li>
<li><a class="" href="dimensionality-reduction-techniques.html"><span class="header-section-number">12</span> Dimensionality Reduction Techniques</a></li>
<li class="book-part">APPENDIX</li>
<li><a class="" href="imap-github-repos.html"><span class="header-section-number">A</span> IMAP GitHub Repos</a></li>
<li><a class="" href="data-preparation-rulegraph.html"><span class="header-section-number">B</span> Data Preparation Rulegraph</a></li>
<li><a class="" href="session-information.html"><span class="header-section-number">C</span> Session Information</a></li>
<li><a class="" href="references.html">References</a></li>
</ul>

        <div class="book-extra">
          <p><a id="book-repo" href="https://github.com/tmbuza/imap-data-preparation.git">View book source <i class="fas fa-air-freshener"></i></a></p>
        </div>
      </nav>
</div>
  </header><main class="col-sm-12 col-md-9 col-lg-7" id="content"><div id="data-tarnsformation" class="section level1" number="9">
<h1>
<span class="header-section-number">9</span> Data Tranformation from Phyloseq Objects<a class="anchor" aria-label="anchor" href="#data-tarnsformation"><i class="fas fa-link"></i></a>
</h1>
<p>The process of data transformation in microbiome analysis involves converting raw or relative abundance values into matrices that are suitable for further analysis. Data transformation focuses on modifying the data itself. In this context, we will explore various methods used for this transformation on a phyloseq object.</p>
<div id="import-processed-phyloseq-object" class="section level2" number="9.1">
<h2>
<span class="header-section-number">9.1</span> Import processed phyloseq object<a class="anchor" aria-label="anchor" href="#import-processed-phyloseq-object"><i class="fas fa-link"></i></a>
</h2>
<div class="sourceCode" id="cb121"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/load.html">load</a></span><span class="op">(</span><span class="st">"../imap-data-preparation/data/phyloseq_raw_rel_psextra_df_objects.rda"</span>, verbose <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> </span></code></pre></div>
<pre><code>Loading objects:
  ps_raw
  ps_rel
  psextra_raw
  psextra_rel
  ps_df</code></pre>
</div>
<div id="raw-abundance" class="section level2" number="9.2">
<h2>
<span class="header-section-number">9.2</span> Raw Abundance<a class="anchor" aria-label="anchor" href="#raw-abundance"><i class="fas fa-link"></i></a>
</h2>
<div class="sourceCode" id="cb123"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ps_raw</span> <span class="op">&lt;-</span> <span class="va">psextra_raw</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/otu_table-methods.html">otu_table</a></span><span class="op">(</span><span class="va">ps_raw</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span></span></code></pre></div>
<pre><code>OTU Table:          [10 taxa and 10 samples]
                     taxa are rows
                            Sample-1 Sample-2 Sample-3 Sample-4 Sample-5
Sporobacter termitidis            87     1187      713      764      138
Eubacterium hallii                10      102       39       31       28
Ruminococcus gnavus               15       56       65       41       47
Klebisiella pneumoniae             3        6        6       22        6
Veillonella                        8       19        6        5        7
Oceanospirillum                    1        2        1        2        1
Ruminococcus lactaris              3        9       23       39        7
Weissella                          1      180        2       13        2
Eggerthella lenta                  3       10        8       18        4
Uncultured Selenomonadaceae        0        0        0        0        0
                            Sample-6 Sample-7 Sample-8 Sample-9 Sample-10
Sporobacter termitidis            81      714      258      159       353
Eubacterium hallii                40       58       26       20        28
Ruminococcus gnavus               61      101       19       21        19
Klebisiella pneumoniae            32        4        5       51        20
Veillonella                        8       11       11       71        74
Oceanospirillum                    1        1        1        1         2
Ruminococcus lactaris             15       22        8        4         4
Weissella                          8        6        1       20         4
Eggerthella lenta                  7       13        6        6         8
Uncultured Selenomonadaceae        0        0        0        0         0</code></pre>
</div>
<div id="no-transformation" class="section level2" number="9.3">
<h2>
<span class="header-section-number">9.3</span> No Transformation<a class="anchor" aria-label="anchor" href="#no-transformation"><i class="fas fa-link"></i></a>
</h2>
<blockquote>
<p>Similar to raw abundance</p>
</blockquote>
<div class="sourceCode" id="cb125"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">(</span><span class="va">ps_identity</span> <span class="op">&lt;-</span> <span class="fu">microbiome</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/microbiome/man/transform.html">transform</a></span><span class="op">(</span><span class="va">ps_raw</span>, <span class="st">'identity'</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code>phyloseq-class experiment-level object
otu_table()   OTU Table:         [ 130 taxa and 222 samples ]
sample_data() Sample Data:       [ 222 samples by 8 sample variables ]
tax_table()   Taxonomy Table:    [ 130 taxa by 3 taxonomic ranks ]</code></pre>
<div class="sourceCode" id="cb127"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/cat.html">cat</a></span><span class="op">(</span><span class="st">"\n\n"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb128"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/otu_table-methods.html">otu_table</a></span><span class="op">(</span><span class="va">ps_identity</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">]</span></span></code></pre></div>
<pre><code>OTU Table:          [5 taxa and 3 samples]
                     taxa are rows
                       Sample-1 Sample-2 Sample-3
Sporobacter termitidis       87     1187      713
Eubacterium hallii           10      102       39
Ruminococcus gnavus          15       56       65
Klebisiella pneumoniae        3        6        6
Veillonella                   8       19        6</code></pre>
</div>
<div id="relative-abundance" class="section level2" number="9.4">
<h2>
<span class="header-section-number">9.4</span> Relative abundance<a class="anchor" aria-label="anchor" href="#relative-abundance"><i class="fas fa-link"></i></a>
</h2>
<div class="sourceCode" id="cb130"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">(</span><span class="va">ps_rel</span> <span class="op">=</span> <span class="fu">phyloseq</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/transformcounts.html">transform_sample_counts</a></span><span class="op">(</span><span class="va">ps_raw</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">{</span><span class="va">x</span> <span class="op">/</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">}</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code>phyloseq-class experiment-level object
otu_table()   OTU Table:         [ 130 taxa and 222 samples ]
sample_data() Sample Data:       [ 222 samples by 8 sample variables ]
tax_table()   Taxonomy Table:    [ 130 taxa by 3 taxonomic ranks ]</code></pre>
<div class="sourceCode" id="cb132"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/otu_table-methods.html">otu_table</a></span><span class="op">(</span><span class="va">ps_rel</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">]</span></span></code></pre></div>
<pre><code>OTU Table:          [5 taxa and 3 samples]
                     taxa are rows
                           Sample-1     Sample-2     Sample-3
Sporobacter termitidis 0.0102849037 0.0562932752 0.0246858013
Eubacterium hallii     0.0011821728 0.0048373328 0.0013502752
Ruminococcus gnavus    0.0017732593 0.0026557906 0.0022504587
Klebisiella pneumoniae 0.0003546519 0.0002845490 0.0002077347
Veillonella            0.0009457383 0.0009010718 0.0002077347</code></pre>
</div>
<div id="arc-sine-asin-transformation" class="section level2" number="9.5">
<h2>
<span class="header-section-number">9.5</span> Arc sine (asin) transformation<a class="anchor" aria-label="anchor" href="#arc-sine-asin-transformation"><i class="fas fa-link"></i></a>
</h2>
<ul>
<li>Typically used when dealing with proportional and percentages.</li>
<li>Proportionals range from 0 to 1</li>
<li>Percentages range from 0 to 100</li>
<li>The Metaphlan3 relative abundances are in percentages! That means the column totals in 100.</li>
</ul>
<div class="sourceCode" id="cb134"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">x</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/otu_table-methods.html">otu_table</a></span><span class="op">(</span><span class="va">ps_rel</span><span class="op">)</span></span>
<span><span class="va">y</span> <span class="op">&lt;-</span> <span class="va">x</span><span class="op">/</span><span class="fu"><a href="https://rdrr.io/r/base/Extremes.html">max</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span></span>
<span></span>
<span><span class="va">ps_asin</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Round.html">round</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Trig.html">asin</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html">sqrt</a></span><span class="op">(</span><span class="va">y</span><span class="op">)</span><span class="op">)</span>, <span class="fl">6</span><span class="op">)</span></span>
<span></span>
<span><span class="va">ps_asin</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span><span class="op">(</span><span class="va">ps_asin</span><span class="op">)</span></span>
<span><span class="va">ps_asin</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">]</span></span></code></pre></div>
<pre><code>OTU Table:          [5 taxa and 4 samples]
                     taxa are rows
                       Sample-1 Sample-2 Sample-3 Sample-4
Sporobacter termitidis 0.115836 0.273804 0.180029 0.277249
Eubacterium hallii     0.039194 0.079347 0.041890 0.055163
Ruminococcus gnavus    0.048009 0.058765 0.054090 0.063450
Klebisiella pneumoniae 0.021464 0.019225 0.016427 0.046464
Veillonella            0.035055 0.034216 0.016427 0.022144</code></pre>
</div>
<div id="compositional-version" class="section level2" number="9.6">
<h2>
<span class="header-section-number">9.6</span> Compositional Version<a class="anchor" aria-label="anchor" href="#compositional-version"><i class="fas fa-link"></i></a>
</h2>
<p>Compositional data represents relative proportions or percentages of different microbial taxa within a sample, rather than the absolute abundance of each taxon. This transformation is necessary because raw abundance data is typically affected by various factors, such as sequencing depth, which can lead to spurious correlations and biases in downstream analyses.</p>
<div class="sourceCode" id="cb136"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">(</span><span class="va">ps_compositional</span> <span class="op">&lt;-</span> <span class="fu">microbiome</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/microbiome/man/transform.html">transform</a></span><span class="op">(</span><span class="va">ps_raw</span>, <span class="st">'compositional'</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code>phyloseq-class experiment-level object
otu_table()   OTU Table:         [ 130 taxa and 222 samples ]
sample_data() Sample Data:       [ 222 samples by 8 sample variables ]
tax_table()   Taxonomy Table:    [ 130 taxa by 3 taxonomic ranks ]</code></pre>
<div class="sourceCode" id="cb138"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/cat.html">cat</a></span><span class="op">(</span><span class="st">"\n\n"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb139"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/otu_table-methods.html">otu_table</a></span><span class="op">(</span><span class="va">ps_compositional</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">]</span></span></code></pre></div>
<pre><code>OTU Table:          [5 taxa and 3 samples]
                     taxa are rows
                           Sample-1     Sample-2     Sample-3
Sporobacter termitidis 0.0102849037 0.0562932752 0.0246858013
Eubacterium hallii     0.0011821728 0.0048373328 0.0013502752
Ruminococcus gnavus    0.0017732593 0.0026557906 0.0022504587
Klebisiella pneumoniae 0.0003546519 0.0002845490 0.0002077347
Veillonella            0.0009457383 0.0009010718 0.0002077347</code></pre>
</div>
<div id="z-transform-for-otus" class="section level2" number="9.7">
<h2>
<span class="header-section-number">9.7</span> Z-transform for OTUs<a class="anchor" aria-label="anchor" href="#z-transform-for-otus"><i class="fas fa-link"></i></a>
</h2>
<div class="sourceCode" id="cb141"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">(</span><span class="va">ps_z_otu</span> <span class="op">&lt;-</span> <span class="fu">microbiome</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/microbiome/man/transform.html">transform</a></span><span class="op">(</span><span class="va">ps_raw</span>, <span class="st">'Z'</span>, <span class="st">'OTU'</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code>phyloseq-class experiment-level object
otu_table()   OTU Table:         [ 130 taxa and 222 samples ]
sample_data() Sample Data:       [ 222 samples by 8 sample variables ]
tax_table()   Taxonomy Table:    [ 130 taxa by 3 taxonomic ranks ]</code></pre>
<div class="sourceCode" id="cb143"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/cat.html">cat</a></span><span class="op">(</span><span class="st">"\n\n"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb144"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/otu_table-methods.html">otu_table</a></span><span class="op">(</span><span class="va">ps_z_otu</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">]</span></span></code></pre></div>
<pre><code>OTU Table:          [5 taxa and 3 samples]
                     taxa are rows
                         Sample-1    Sample-2    Sample-3
Sporobacter termitidis -0.7794277  1.90302575  1.37827866
Eubacterium hallii     -0.4477855  2.45470119  1.22737899
Ruminococcus gnavus    -0.4074006  1.29211010  1.48822307
Klebisiella pneumoniae -0.7431343 -0.01469595 -0.01469595
Veillonella             0.6448912  1.68310449  0.31813426</code></pre>
</div>
<div id="z-transform-for-samples" class="section level2" number="9.8">
<h2>
<span class="header-section-number">9.8</span> Z-transform for Samples<a class="anchor" aria-label="anchor" href="#z-transform-for-samples"><i class="fas fa-link"></i></a>
</h2>
<div class="sourceCode" id="cb146"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">(</span><span class="va">ps_z_sample</span> <span class="op">&lt;-</span> <span class="fu">microbiome</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/microbiome/man/transform.html">transform</a></span><span class="op">(</span><span class="va">ps_raw</span>, <span class="st">'Z'</span>, <span class="st">'sample'</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code>phyloseq-class experiment-level object
otu_table()   OTU Table:         [ 130 taxa and 222 samples ]
sample_data() Sample Data:       [ 222 samples by 8 sample variables ]
tax_table()   Taxonomy Table:    [ 130 taxa by 3 taxonomic ranks ]</code></pre>
<div class="sourceCode" id="cb148"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/cat.html">cat</a></span><span class="op">(</span><span class="st">"\n\n"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb149"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/otu_table-methods.html">otu_table</a></span><span class="op">(</span><span class="va">ps_z_sample</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">]</span></span></code></pre></div>
<pre><code>OTU Table:          [5 taxa and 3 samples]
                     taxa are rows
                          Sample-1   Sample-2   Sample-3
Sporobacter termitidis  1.23557381  2.0958437  1.7796631
Eubacterium hallii      0.12679267  0.9517496  0.4813434
Ruminococcus gnavus     0.32658334  0.6749183  0.7069387
Klebisiella pneumoniae -0.41260408 -0.3062820 -0.3038504
Veillonella             0.01979284  0.1849037 -0.3038504</code></pre>
</div>
<div id="log10-transform" class="section level2" number="9.9">
<h2>
<span class="header-section-number">9.9</span> Log10 Transform<a class="anchor" aria-label="anchor" href="#log10-transform"><i class="fas fa-link"></i></a>
</h2>
<div class="sourceCode" id="cb151"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">(</span><span class="va">ps_log10</span> <span class="op">&lt;-</span> <span class="fu">microbiome</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/microbiome/man/transform.html">transform</a></span><span class="op">(</span><span class="va">ps_raw</span>, <span class="st">'log10'</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code>phyloseq-class experiment-level object
otu_table()   OTU Table:         [ 130 taxa and 222 samples ]
sample_data() Sample Data:       [ 222 samples by 8 sample variables ]
tax_table()   Taxonomy Table:    [ 130 taxa by 3 taxonomic ranks ]</code></pre>
<div class="sourceCode" id="cb153"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/cat.html">cat</a></span><span class="op">(</span><span class="st">"\n\n"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb154"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/otu_table-methods.html">otu_table</a></span><span class="op">(</span><span class="va">ps_log10</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">]</span></span></code></pre></div>
<pre><code>OTU Table:          [5 taxa and 3 samples]
                     taxa are rows
                        Sample-1 Sample-2 Sample-3
Sporobacter termitidis 1.9444827 3.074816 2.853698
Eubacterium hallii     1.0413927 2.012837 1.602060
Ruminococcus gnavus    1.2041200 1.755875 1.819544
Klebisiella pneumoniae 0.6020600 0.845098 0.845098
Veillonella            0.9542425 1.301030 0.845098</code></pre>
</div>
<div id="log10p-transform" class="section level2" number="9.10">
<h2>
<span class="header-section-number">9.10</span> Log10p Transform<a class="anchor" aria-label="anchor" href="#log10p-transform"><i class="fas fa-link"></i></a>
</h2>
<div class="sourceCode" id="cb156"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">(</span><span class="va">ps_log10p</span> <span class="op">&lt;-</span> <span class="fu">microbiome</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/microbiome/man/transform.html">transform</a></span><span class="op">(</span><span class="va">ps_raw</span>, <span class="st">'log10p'</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code>phyloseq-class experiment-level object
otu_table()   OTU Table:         [ 130 taxa and 222 samples ]
sample_data() Sample Data:       [ 222 samples by 8 sample variables ]
tax_table()   Taxonomy Table:    [ 130 taxa by 3 taxonomic ranks ]</code></pre>
<div class="sourceCode" id="cb158"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/cat.html">cat</a></span><span class="op">(</span><span class="st">"\n\n"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb159"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/otu_table-methods.html">otu_table</a></span><span class="op">(</span><span class="va">ps_log10p</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">]</span></span></code></pre></div>
<pre><code>OTU Table:          [5 taxa and 3 samples]
                     taxa are rows
                        Sample-1 Sample-2 Sample-3
Sporobacter termitidis 1.9444827 3.074816 2.853698
Eubacterium hallii     1.0413927 2.012837 1.602060
Ruminococcus gnavus    1.2041200 1.755875 1.819544
Klebisiella pneumoniae 0.6020600 0.845098 0.845098
Veillonella            0.9542425 1.301030 0.845098</code></pre>
</div>
<div id="clr-transform" class="section level2" number="9.11">
<h2>
<span class="header-section-number">9.11</span> CLR Transform<a class="anchor" aria-label="anchor" href="#clr-transform"><i class="fas fa-link"></i></a>
</h2>
<ul>
<li>Note that small pseudocount is added if data contains zeroes</li>
</ul>
<div class="sourceCode" id="cb161"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">(</span><span class="va">ps_clr</span> <span class="op">&lt;-</span> <span class="fu">microbiome</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/microbiome/man/transform.html">transform</a></span><span class="op">(</span><span class="va">ps_raw</span>, <span class="st">'clr'</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code>phyloseq-class experiment-level object
otu_table()   OTU Table:         [ 130 taxa and 222 samples ]
sample_data() Sample Data:       [ 222 samples by 8 sample variables ]
tax_table()   Taxonomy Table:    [ 130 taxa by 3 taxonomic ranks ]</code></pre>
<div class="sourceCode" id="cb163"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/cat.html">cat</a></span><span class="op">(</span><span class="st">"\n\n"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb164"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/otu_table-methods.html">otu_table</a></span><span class="op">(</span><span class="va">ps_clr</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">]</span></span></code></pre></div>
<pre><code>OTU Table:          [5 taxa and 3 samples]
                     taxa are rows
                         Sample-1   Sample-2  Sample-3
Sporobacter termitidis  2.8332848  4.7278340  4.148346
Eubacterium hallii      0.6828174  2.2768872  1.254464
Ruminococcus gnavus     1.0834602  1.6801909  1.760214
Klebisiella pneumoniae -0.4880348 -0.5008398 -0.550034
Veillonella             0.4632754  0.6118107 -0.550034</code></pre>
</div>
<div id="shift-the-baseline" class="section level2" number="9.12">
<h2>
<span class="header-section-number">9.12</span> Shift the baseline<a class="anchor" aria-label="anchor" href="#shift-the-baseline"><i class="fas fa-link"></i></a>
</h2>
<div class="sourceCode" id="cb166"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">(</span><span class="va">ps_shift</span> <span class="op">&lt;-</span> <span class="fu">microbiome</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/microbiome/man/transform.html">transform</a></span><span class="op">(</span><span class="va">ps_raw</span>, <span class="st">'shift'</span>, shift<span class="op">=</span><span class="fl">1</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code>phyloseq-class experiment-level object
otu_table()   OTU Table:         [ 130 taxa and 222 samples ]
sample_data() Sample Data:       [ 222 samples by 8 sample variables ]
tax_table()   Taxonomy Table:    [ 130 taxa by 3 taxonomic ranks ]</code></pre>
<div class="sourceCode" id="cb168"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/cat.html">cat</a></span><span class="op">(</span><span class="st">"\n\n"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb169"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/otu_table-methods.html">otu_table</a></span><span class="op">(</span><span class="va">ps_shift</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">]</span></span></code></pre></div>
<pre><code>OTU Table:          [5 taxa and 3 samples]
                     taxa are rows
                       Sample-1 Sample-2 Sample-3
Sporobacter termitidis       88     1188      714
Eubacterium hallii           11      103       40
Ruminococcus gnavus          16       57       66
Klebisiella pneumoniae        4        7        7
Veillonella                   9       20        7</code></pre>
</div>
<div id="data-scaling" class="section level2" number="9.13">
<h2>
<span class="header-section-number">9.13</span> Data Scaling<a class="anchor" aria-label="anchor" href="#data-scaling"><i class="fas fa-link"></i></a>
</h2>
<div class="sourceCode" id="cb171"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">(</span><span class="va">ps_scale</span> <span class="op">&lt;-</span> <span class="fu">microbiome</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/microbiome/man/transform.html">transform</a></span><span class="op">(</span><span class="va">ps_raw</span>, <span class="st">'scale'</span>, scale<span class="op">=</span><span class="fl">1</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code>phyloseq-class experiment-level object
otu_table()   OTU Table:         [ 130 taxa and 222 samples ]
sample_data() Sample Data:       [ 222 samples by 8 sample variables ]
tax_table()   Taxonomy Table:    [ 130 taxa by 3 taxonomic ranks ]</code></pre>
<div class="sourceCode" id="cb173"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/cat.html">cat</a></span><span class="op">(</span><span class="st">"\n\n"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb174"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/otu_table-methods.html">otu_table</a></span><span class="op">(</span><span class="va">ps_scale</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">]</span></span></code></pre></div>
<pre><code>OTU Table:          [5 taxa and 3 samples]
                     taxa are rows
                       Sample-1 Sample-2 Sample-3
Sporobacter termitidis       87     1187      713
Eubacterium hallii           10      102       39
Ruminococcus gnavus          15       56       65
Klebisiella pneumoniae        3        6        6
Veillonella                   8       19        6</code></pre>
</div>
<div id="save-transformed-objects" class="section level2" number="9.14">
<h2>
<span class="header-section-number">9.14</span> Save transformed objects<a class="anchor" aria-label="anchor" href="#save-transformed-objects"><i class="fas fa-link"></i></a>
</h2>
<div class="sourceCode" id="cb176"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/save.html">save</a></span><span class="op">(</span></span>
<span>  <span class="va">ps_asin</span>, </span>
<span>  <span class="va">ps_identity</span>,</span>
<span>  <span class="va">ps_compositional</span>, </span>
<span>  <span class="va">ps_z_otu</span>, </span>
<span>  <span class="va">ps_z_sample</span>, </span>
<span>  <span class="va">ps_log10</span>, </span>
<span>  <span class="va">ps_log10p</span>, </span>
<span>  <span class="va">ps_clr</span>, </span>
<span>  <span class="va">ps_shift</span>, </span>
<span>  <span class="va">ps_scale</span>, </span>
<span>  file <span class="op">=</span> <span class="st">"data/ps_transformed.rda"</span><span class="op">)</span></span></code></pre></div>
</div>
<div id="confirm-saved-object" class="section level2" number="9.15">
<h2>
<span class="header-section-number">9.15</span> Confirm saved object<a class="anchor" aria-label="anchor" href="#confirm-saved-object"><i class="fas fa-link"></i></a>
</h2>
<div class="sourceCode" id="cb177"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/load.html">load</a></span><span class="op">(</span><span class="st">"data/ps_transformed.rda"</span>, verbose <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<pre><code>Loading objects:
  ps_asin
  ps_identity
  ps_compositional
  ps_z_otu
  ps_z_sample
  ps_log10
  ps_log10p
  ps_clr
  ps_shift
  ps_scale</code></pre>

</div>
</div>



  <div class="chapter-nav">
<div class="prev"><a href="pipeline-for-cleaning-phyloseq-objects.html"><span class="header-section-number">8</span> Pipeline for Cleaning Phyloseq Objects</a></div>
<div class="next"><a href="distance-calculation-in-microbiome-data-analysis.html"><span class="header-section-number">10</span> Distance Calculation in Microbiome Data Analysis</a></div>
</div></main><div class="col-md-3 col-lg-2 d-none d-md-block sidebar sidebar-chapter">
    <nav id="toc" data-toggle="toc" aria-label="On this page"><h2>On this page</h2>
      <ul class="nav navbar-nav">
<li><a class="nav-link" href="#data-tarnsformation"><span class="header-section-number">9</span> Data Tranformation from Phyloseq Objects</a></li>
<li><a class="nav-link" href="#import-processed-phyloseq-object"><span class="header-section-number">9.1</span> Import processed phyloseq object</a></li>
<li><a class="nav-link" href="#raw-abundance"><span class="header-section-number">9.2</span> Raw Abundance</a></li>
<li><a class="nav-link" href="#no-transformation"><span class="header-section-number">9.3</span> No Transformation</a></li>
<li><a class="nav-link" href="#relative-abundance"><span class="header-section-number">9.4</span> Relative abundance</a></li>
<li><a class="nav-link" href="#arc-sine-asin-transformation"><span class="header-section-number">9.5</span> Arc sine (asin) transformation</a></li>
<li><a class="nav-link" href="#compositional-version"><span class="header-section-number">9.6</span> Compositional Version</a></li>
<li><a class="nav-link" href="#z-transform-for-otus"><span class="header-section-number">9.7</span> Z-transform for OTUs</a></li>
<li><a class="nav-link" href="#z-transform-for-samples"><span class="header-section-number">9.8</span> Z-transform for Samples</a></li>
<li><a class="nav-link" href="#log10-transform"><span class="header-section-number">9.9</span> Log10 Transform</a></li>
<li><a class="nav-link" href="#log10p-transform"><span class="header-section-number">9.10</span> Log10p Transform</a></li>
<li><a class="nav-link" href="#clr-transform"><span class="header-section-number">9.11</span> CLR Transform</a></li>
<li><a class="nav-link" href="#shift-the-baseline"><span class="header-section-number">9.12</span> Shift the baseline</a></li>
<li><a class="nav-link" href="#data-scaling"><span class="header-section-number">9.13</span> Data Scaling</a></li>
<li><a class="nav-link" href="#save-transformed-objects"><span class="header-section-number">9.14</span> Save transformed objects</a></li>
<li><a class="nav-link" href="#confirm-saved-object"><span class="header-section-number">9.15</span> Confirm saved object</a></li>
</ul>

      <div class="book-extra">
        <ul class="list-unstyled">
<li><a id="book-source" href="https://github.com/tmbuza/imap-data-preparation.git/blob/gh-pages/05_data_transformation.Rmd">View source <i class="fas fa-air-freshener"></i></a></li>
          <li><a id="book-edit" href="https://github.com/tmbuza/imap-data-preparation.git/edit/gh-pages/05_data_transformation.Rmd">Edit this page <i class="fas fa-air-freshener"></i></a></li>
        </ul>
</div>
    </nav>
</div>

</div>
</div> <!-- .container -->

<footer class="bg-primary text-light mt-5"><div class="container">
    <div class="row">
      <div class="col-12 col-md-6 mt-3">
        <p><strong><u><a class="text-light" href="https://microbiome.complexdatainsights.com"><big>IMAP:</big> Integrated Microbiome Analysis Pipelines</a></u></strong>: End-to-End Practical User Guides Using Integrated Approaches</p>
<br><p>Authored by Teresia Mrema-Buza, Last updated: <b>2024-04-17</b>.</p>
<br>
</div>
      <div class="col-12 col-md-6 mt-3">
        <p>Note: This practical guide was created using the <a class="text-light" href="https://bookdown.org/yihui/bookdown/">bookdown</a> R package, styled with the modern bs4_book format.</p>
      </div>

    </div>
  </div>
</footer>
</body>
</html>
