<!DOCTYPE html>
<html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>4 Import Preprocessed Microbiome Data | IMAP: Integrated Microbiome Analysis Pipelines</title>
<meta name="author" content="Teresia Mrema-Buza">
<meta name="description" content="In this section, we are presenting the process of importing authentic preprocessed data, which will serve as the cornerstone for our subsequent analyses. Although synthetic data certainly has its...">
<meta name="generator" content="bookdown 0.37 with bs4_book()">
<meta property="og:title" content="4 Import Preprocessed Microbiome Data | IMAP: Integrated Microbiome Analysis Pipelines">
<meta property="og:type" content="book">
<meta property="og:url" content="https://tmbuza.github.io/imap-data-preparation/import-preprocessed-microbiome-data.html">
<meta property="og:image" content="https://tmbuza.github.io/imap-data-preparation/images/planning.png">
<meta property="og:description" content="In this section, we are presenting the process of importing authentic preprocessed data, which will serve as the cornerstone for our subsequent analyses. Although synthetic data certainly has its...">
<meta name="twitter:card" content="summary">
<meta name="twitter:title" content="4 Import Preprocessed Microbiome Data | IMAP: Integrated Microbiome Analysis Pipelines">
<meta name="twitter:description" content="In this section, we are presenting the process of importing authentic preprocessed data, which will serve as the cornerstone for our subsequent analyses. Although synthetic data certainly has its...">
<meta name="twitter:image" content="https://tmbuza.github.io/imap-data-preparation/images/planning.png">
<!-- JS --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://kit.fontawesome.com/6ecbd6c532.js" crossorigin="anonymous"></script><script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="libs/bootstrap-4.6.0/bootstrap.min.css" rel="stylesheet">
<script src="libs/bootstrap-4.6.0/bootstrap.bundle.min.js"></script><link href="libs/Sen-0.4.8/font.css" rel="stylesheet">
<link href="https://fonts.googleapis.com/css2?family=DM%20Mono&amp;display=swap" rel="stylesheet">
<link href="libs/Bitter-0.4.8/font.css" rel="stylesheet">
<script src="libs/bs3compat-0.6.1/transition.js"></script><script src="libs/bs3compat-0.6.1/tabs.js"></script><script src="libs/bs3compat-0.6.1/bs3compat.js"></script><link href="libs/bs4_book-1.0.0/bs4_book.css" rel="stylesheet">
<script src="libs/bs4_book-1.0.0/bs4_book.js"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- CSS --><style type="text/css">
    
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  </style>
<link rel="stylesheet" href="style.css">
</head>
<body data-spy="scroll" data-target="#toc">

<div class="container-fluid">
<div class="row">
  <header class="col-sm-12 col-lg-3 sidebar sidebar-book"><a class="sr-only sr-only-focusable" href="#content">Skip to main content</a>

    <div class="d-flex align-items-start justify-content-between">
      <h1>
        <a href="index.html" title="End-to-End Practical User Guides Using Integrated Approaches"><big>IMAP:</big> Integrated Microbiome Analysis Pipelines</a>:
        <small class="text-muted">End-to-End Practical User Guides Using Integrated Approaches</small>
      </h1>
      <button class="btn btn-outline-primary d-lg-none ml-2 mt-1" type="button" data-toggle="collapse" data-target="#main-nav" aria-expanded="true" aria-controls="main-nav"><i class="fas fa-bars"></i><span class="sr-only">Show table of contents</span></button>
    </div>

    <div id="main-nav" class="collapse-lg">
      <form role="search">
        <input id="search" class="form-control" type="search" placeholder="Search" aria-label="Search">
</form>

      <nav aria-label="Table of contents"><h2>Table of contents</h2>
        <ul class="book-toc list-unstyled">
<li><a class="" href="index.html">IMAP-Part 07: Microbiome Data Preparation Guide</a></li>
<li class="book-part">GETTING STARTED</li>
<li><a class="" href="preface.html">Preparing for Downstream Analysis</a></li>
<li><a class="" href="data-tidying.html"><span class="header-section-number">1</span> Data Tidying</a></li>
<li class="book-part">MOTHUR OUTPUT</li>
<li><a class="" href="prepare-mothur-output.html"><span class="header-section-number">2</span> Prepare Mothur Output</a></li>
<li class="book-part">QIIME2 OUTPUT</li>
<li><a class="" href="prepare-qiime2-output.html"><span class="header-section-number">3</span> Prepare Qiime2 Output</a></li>
<li class="book-part">EXTERNAL DATASETS</li>
<li><a class="active" href="import-preprocessed-microbiome-data.html"><span class="header-section-number">4</span> Import Preprocessed Microbiome Data</a></li>
<li class="book-part">DATA TRANSFORMATION</li>
<li><a class="" href="data-tarnsformation.html"><span class="header-section-number">5</span> Data Tranformation from Phyloseq Objects</a></li>
<li class="book-part">DIMENSIONAL REDUCTION</li>
<li><a class="" href="data-reduction.html"><span class="header-section-number">6</span> Dimensionality Reduction</a></li>
<li class="book-part">APPENDIX</li>
<li><a class="" href="imap-github-repos.html"><span class="header-section-number">A</span> IMAP GitHub Repos</a></li>
<li><a class="" href="data-preparation-rulegraph.html"><span class="header-section-number">B</span> Data Preparation Rulegraph</a></li>
<li><a class="" href="session-information.html"><span class="header-section-number">C</span> Session Information</a></li>
<li><a class="" href="references.html">References</a></li>
</ul>

        <div class="book-extra">
          <p><a id="book-repo" href="https://github.com/tmbuza/imap-data-preparation.git">View book source <i class="fas fa-air-freshener"></i></a></p>
        </div>
      </nav>
</div>
  </header><main class="col-sm-12 col-md-9 col-lg-7" id="content"><div id="import-preprocessed-microbiome-data" class="section level1" number="4">
<h1>
<span class="header-section-number">4</span> Import Preprocessed Microbiome Data<a class="anchor" aria-label="anchor" href="#import-preprocessed-microbiome-data"><i class="fas fa-link"></i></a>
</h1>
<p>In this section, we are presenting the process of importing authentic preprocessed data, which will serve as the cornerstone for our subsequent analyses. Although synthetic data certainly has its advantages, it can at times introduce biases or limitations that do not fully capture the intricate diversity inherent in real microbiome samples.</p>
<div id="the-globalpatterns-dataset" class="section level2" number="4.1">
<h2>
<span class="header-section-number">4.1</span> The GlobalPatterns dataset<a class="anchor" aria-label="anchor" href="#the-globalpatterns-dataset"><i class="fas fa-link"></i></a>
</h2>
<p>The GlobalPatterns dataset, sourced from the Earth Microbiome Project (EMP), serves as a comprehensive repository for studying microbial communities worldwide. Here’s a breakdown of its key attributes:</p>
<ul>
<li><p><strong>Source:</strong> GlobalPatterns originates from the Earth Microbiome Project (EMP), collecting samples worldwide.</p></li>
<li><p><strong>Composition:</strong> It comprises high-throughput sequencing data, revealing the taxonomic composition of microbial communities.</p></li>
<li><p><strong>Scope:</strong> Samples represent diverse global ecosystems, offering a comprehensive view of microbial biodiversity.</p></li>
<li><p><strong>Format:</strong> Presented as a phyloseq object in R, it integrates sample metadata and taxonomic abundance for analysis.</p></li>
<li><p><strong>Utility:</strong> Researchers utilize it for community profiling, abundance testing, and ecological modeling, enhancing understanding of global microbial diversity and function.</p></li>
</ul>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://dx.plos.org/10.1371/journal.pone.0061217">phyloseq</a></span><span class="op">)</span> <span class="co"># for GlobalPatterns dataset</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="st">"GlobalPatterns"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">ps_GlobalPatterns</span> <span class="op">&lt;-</span><span class="va">GlobalPatterns</span></span>
<span><span class="va">df_GlobalPatterns</span> <span class="op">&lt;-</span><span class="va">GlobalPatterns</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu">phyloseq</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/psmelt.html">psmelt</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu">tibble</span><span class="fu">::</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/rownames.html">rownames_to_column</a></span><span class="op">(</span><span class="st">"sample_id"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select_all.html">rename_all</a></span><span class="op">(</span><span class="va">tolower</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">df_GlobalPatterns</span><span class="op">)</span></span></code></pre></div>
<pre><code> [1] "sample_id"                "otu"                     
 [3] "sample"                   "abundance"               
 [5] "x.sampleid"               "primer"                  
 [7] "final_barcode"            "barcode_truncated_plus_t"
 [9] "barcode_full_length"      "sampletype"              
[11] "description"              "kingdom"                 
[13] "phylum"                   "class"                   
[15] "order"                    "family"                  
[17] "genus"                    "species"                 </code></pre>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/cat.html">cat</a></span><span class="op">(</span><span class="st">"\n"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ps_GlobalPatterns</span></span></code></pre></div>
<pre><code>phyloseq-class experiment-level object
otu_table()   OTU Table:         [ 19216 taxa and 26 samples ]
sample_data() Sample Data:       [ 26 samples by 7 sample variables ]
tax_table()   Taxonomy Table:    [ 19216 taxa by 7 taxonomic ranks ]
phy_tree()    Phylogenetic Tree: [ 19216 tips and 19215 internal nodes ]</code></pre>
</div>
<div id="the-dietswap-dataset" class="section level2" number="4.2">
<h2>
<span class="header-section-number">4.2</span> The dietswap dataset<a class="anchor" aria-label="anchor" href="#the-dietswap-dataset"><i class="fas fa-link"></i></a>
</h2>
<p>The dietswap dataset, available through the microbiome R package, offers valuable insights into the effects of dietary interventions on microbial communities. Here’s an overview of its attributes:</p>
<ul>
<li><p><strong>Source:</strong> The dietswap dataset is derived from research examining the impact of dietary changes on the human gut microbiome.</p></li>
<li><p><strong>Composition:</strong> It comprises high-throughput sequencing data, providing insights into the taxonomic composition and dynamics of microbial communities in response to dietary alterations.</p></li>
<li><p><strong>Scope:</strong> Samples are obtained from human participants undergoing dietary interventions, enabling researchers to explore how different diets influence microbial diversity and function within the gut microbiome.</p></li>
<li><p><strong>Format:</strong> Presented in a format suitable for microbiome analysis, the dataset includes sample metadata and taxonomic abundance data, facilitating comprehensive analyses of microbial community dynamics.</p></li>
<li><p><strong>Utility:</strong> Researchers utilize the dietswap dataset to investigate the effects of dietary interventions on gut microbiome composition, contributing to a better understanding of the intricate interactions between diet, host physiology, and microbial ecology.</p></li>
</ul>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://microbiome.github.io/microbiome">microbiome</a></span><span class="op">)</span> <span class="co"># for dietswap dataset</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="st">"dietswap"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">ps_dietswap</span> <span class="op">&lt;-</span><span class="va">dietswap</span></span>
<span><span class="va">df_dietswap</span> <span class="op">&lt;-</span><span class="va">dietswap</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu">phyloseq</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/psmelt.html">psmelt</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu">tibble</span><span class="fu">::</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/rownames.html">rownames_to_column</a></span><span class="op">(</span><span class="st">"sample_id"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span><span class="op">-</span><span class="fl">9</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select_all.html">rename_all</a></span><span class="op">(</span><span class="va">tolower</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">df_dietswap</span><span class="op">)</span></span></code></pre></div>
<pre><code> [1] "sample_id"              "otu"                    "sample"                
 [4] "abundance"              "subject"                "sex"                   
 [7] "nationality"            "group"                  "timepoint"             
[10] "timepoint.within.group" "bmi_group"              "phylum"                
[13] "family"                 "genus"                 </code></pre>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/cat.html">cat</a></span><span class="op">(</span><span class="st">"\n"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ps_dietswap</span></span></code></pre></div>
<pre><code>phyloseq-class experiment-level object
otu_table()   OTU Table:         [ 130 taxa and 222 samples ]
sample_data() Sample Data:       [ 222 samples by 8 sample variables ]
tax_table()   Taxonomy Table:    [ 130 taxa by 3 taxonomic ranks ]</code></pre>
</div>
<div id="caporaso-dataset" class="section level2" number="4.3">
<h2>
<span class="header-section-number">4.3</span> Caporaso dataset<a class="anchor" aria-label="anchor" href="#caporaso-dataset"><i class="fas fa-link"></i></a>
</h2>
<p>The Caporaso dataset provides valuable insights into microbial marker genes and their associations with various biological factors.</p>
<ul>
<li><p><strong>Source:</strong> The Caporaso dataset is included in the <code>microbiomeMarker</code> R package and is derived from research led by Dr. J. Gregory Caporaso. It aims to provide insights into microbial marker genes and their associations with various biological factors.</p></li>
<li><p><strong>Composition:</strong> The dataset comprises high-throughput sequencing data, focusing on microbial marker genes from diverse biological samples. It offers valuable information regarding the taxonomic composition and functional potential of microbial communities.</p></li>
<li><p><strong>Scope:</strong> Samples are collected from a range of environments, including but not limited to human microbiomes, environmental samples, and animal microbiomes. This diversity allows researchers to explore microbial diversity across different ecosystems and conditions.</p></li>
<li><p><strong>Format:</strong> The dataset is structured to facilitate microbiome marker analysis, with sample metadata and taxonomic abundance data included. This format enables researchers to conduct comprehensive analyses of microbial marker genes and their associations with environmental or biological factors.</p></li>
<li><p><strong>Utility:</strong> Researchers utilize the Caporaso dataset to investigate microbial marker genes’ roles in various ecosystems, such as host-microbiome interactions, environmental responses, and disease states. The dataset contributes to a better understanding of microbial ecology and its implications for human health and environmental management.</p></li>
</ul>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/yiluheihei/microbiomeMarker">microbiomeMarker</a></span><span class="op">)</span> <span class="co"># for caporaso dataset</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="st">"caporaso"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">ps_caporaso</span> <span class="op">&lt;-</span><span class="va">caporaso</span></span>
<span><span class="va">df_caporaso</span> <span class="op">&lt;-</span><span class="va">caporaso</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu">phyloseq</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/psmelt.html">psmelt</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu">tibble</span><span class="fu">::</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/rownames.html">rownames_to_column</a></span><span class="op">(</span><span class="st">"sample_id"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select_all.html">rename_all</a></span><span class="op">(</span><span class="va">tolower</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">df_caporaso</span><span class="op">)</span></span></code></pre></div>
<pre><code> [1] "sample_id"                "otu"                     
 [3] "sample"                   "abundance"               
 [5] "sampletype"               "year"                    
 [7] "month"                    "day"                     
 [9] "subject"                  "reportedantibioticusage" 
[11] "dayssinceexperimentstart" "description"             
[13] "kingdom"                  "phylum"                  
[15] "class"                    "order"                   
[17] "family"                   "genus"                   
[19] "species"                 </code></pre>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/cat.html">cat</a></span><span class="op">(</span><span class="st">"\n"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ps_caporaso</span></span></code></pre></div>
<pre><code>phyloseq-class experiment-level object
otu_table()   OTU Table:         [ 3426 taxa and 34 samples ]
sample_data() Sample Data:       [ 34 samples by 8 sample variables ]
tax_table()   Taxonomy Table:    [ 3426 taxa by 7 taxonomic ranks ]
phy_tree()    Phylogenetic Tree: [ 3426 tips and 3424 internal nodes ]</code></pre>
</div>
<div id="kostic_crc-dataset" class="section level2" number="4.4">
<h2>
<span class="header-section-number">4.4</span> Kostic_CRC dataset<a class="anchor" aria-label="anchor" href="#kostic_crc-dataset"><i class="fas fa-link"></i></a>
</h2>
<p>The Kostic_CRC dataset provides valuable insights into the gut microbiome of individuals diagnosed with colorectal cancer (CRC).</p>
<ul>
<li><p><strong>Source:</strong> The Kostic_CRC dataset is included in the <code>microbiomeMarker</code> R package. It is derived from research conducted by the Kostic Lab and focuses on investigating the gut microbiome in colorectal cancer (CRC) patients.</p></li>
<li><p><strong>Composition:</strong> This dataset comprises high-throughput sequencing data, specifically targeting the gut microbiome of individuals with colorectal cancer. It provides insights into the taxonomic composition and potential functional characteristics of microbial communities associated with CRC.</p></li>
<li><p><strong>Scope:</strong> Samples are collected from individuals diagnosed with colorectal cancer, allowing researchers to explore the microbial diversity and potential biomarkers associated with CRC development and progression.</p></li>
<li><p><strong>Format:</strong> The dataset is structured to facilitate microbiome marker analysis, including sample metadata and taxonomic abundance data. This format enables researchers to conduct comprehensive analyses of microbial community dynamics in colorectal cancer.</p></li>
<li><p><strong>Utility:</strong> Researchers utilize the Kostic_CRC dataset to investigate the role of the gut microbiome in colorectal cancer pathogenesis, prognosis, and treatment response. The dataset contributes to a deeper understanding of the complex interplay between the gut microbiome and colorectal cancer biology, potentially leading to novel diagnostic or therapeutic strategies.</p></li>
</ul>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/yiluheihei/microbiomeMarker">microbiomeMarker</a></span><span class="op">)</span> <span class="co"># for kostic_crc dataset</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="st">"kostic_crc"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">ps_kostic_crc</span> <span class="op">&lt;-</span><span class="va">kostic_crc</span></span>
<span><span class="va">df_kostic_crc</span> <span class="op">&lt;-</span><span class="va">kostic_crc</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu">phyloseq</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/psmelt.html">psmelt</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu">tibble</span><span class="fu">::</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/rownames.html">rownames_to_column</a></span><span class="op">(</span><span class="st">"sample_id"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select_all.html">rename_all</a></span><span class="op">(</span><span class="va">tolower</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">df_kostic_crc</span><span class="op">)</span></span></code></pre></div>
<pre><code> [1] "sample_id"                     "otu"                          
 [3] "sample"                        "abundance"                    
 [5] "x.sampleid"                    "barcodesequence"              
 [7] "linkerprimersequence"          "necrosis_percent"             
 [9] "target_subfragment"            "assigned_from_geo"            
[11] "experiment_center"             "title"                        
[13] "run_prefix"                    "age"                          
[15] "normal_equivalent_percent"     "fibroblast_and_vessel_percent"
[17] "depth"                         "treatment"                    
[19] "age_at_diagnosis"              "common_name"                  
[21] "host_common_name"              "body_site"                    
[23] "elevation"                     "reports_received"             
[25] "cea"                           "pcr_primers"                  
[27] "collection_date"               "altitude"                     
[29] "env_biome"                     "sex"                          
[31] "platform"                      "race"                         
[33] "bsp_diagnosis"                 "study_center"                 
[35] "country"                       "chemotherapy"                 
[37] "year_of_death"                 "ethnicity"                    
[39] "anonymized_name"               "taxon_id"                     
[41] "sample_center"                 "samp_size"                    
[43] "year_of_birth"                 "original_diagnosis"           
[45] "age_unit"                      "study_id"                     
[47] "experiment_design_description" "description_duplicate"        
[49] "diagnosis"                     "body_habitat"                 
[51] "sequencing_meth"               "run_date"                     
[53] "histologic_grade"              "longitude"                    
[55] "env_matter"                    "target_gene"                  
[57] "env_feature"                   "key_seq"                      
[59] "body_product"                  "tumor_percent"                
[61] "library_construction_protocol" "region"                       
[63] "run_center"                    "tumor_type"                   
[65] "bsp_notes"                     "radiation_therapy"            
[67] "inflammation_percent"          "host_subject_id"              
[69] "pc3"                           "latitude"                     
[71] "osh_diagnosis"                 "stage"                        
[73] "primary_disease"               "host_taxid"                   
[75] "description"                   "kingdom"                      
[77] "phylum"                        "class"                        
[79] "order"                         "family"                       
[81] "genus"                         "species"                      </code></pre>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/cat.html">cat</a></span><span class="op">(</span><span class="st">"\n"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ps_kostic_crc</span></span></code></pre></div>
<pre><code>phyloseq-class experiment-level object
otu_table()   OTU Table:         [ 2505 taxa and 177 samples ]
sample_data() Sample Data:       [ 177 samples by 71 sample variables ]
tax_table()   Taxonomy Table:    [ 2505 taxa by 7 taxonomic ranks ]</code></pre>
</div>
<div id="save_ps_objects" class="section level2" number="4.5">
<h2>
<span class="header-section-number">4.5</span> Save objects for transformation and exploration<a class="anchor" aria-label="anchor" href="#save_ps_objects"><i class="fas fa-link"></i></a>
</h2>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/save.html">save</a></span><span class="op">(</span><span class="va">df_GlobalPatterns</span>, </span>
<span>     <span class="co"># df_ibd_phylo, </span></span>
<span>     <span class="va">df_dietswap</span>,  </span>
<span>     <span class="va">df_caporaso</span>,</span>
<span>     <span class="va">df_kostic_crc</span>,</span>
<span></span>
<span>     <span class="va">ps_GlobalPatterns</span>, </span>
<span>     <span class="co"># ps_ibd_phylo, </span></span>
<span>     <span class="va">ps_dietswap</span>,</span>
<span>     <span class="va">ps_caporaso</span>,</span>
<span>     <span class="va">ps_kostic_crc</span>,</span>
<span>     file <span class="op">=</span> <span class="st">"data/external/external_ps_objects.rda"</span><span class="op">)</span></span></code></pre></div>
</div>
<div id="confirm-saved-objects" class="section level2" number="4.6">
<h2>
<span class="header-section-number">4.6</span> Confirm saved objects<a class="anchor" aria-label="anchor" href="#confirm-saved-objects"><i class="fas fa-link"></i></a>
</h2>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/load.html">load</a></span><span class="op">(</span><span class="st">"data/external/external_ps_objects.rda"</span>, verbose <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<pre><code>Loading objects:
  df_GlobalPatterns
  df_dietswap
  df_caporaso
  df_kostic_crc
  ps_GlobalPatterns
  ps_dietswap
  ps_caporaso
  ps_kostic_crc</code></pre>
</div>
<div id="review-phyloseq-objects" class="section level2" number="4.7">
<h2>
<span class="header-section-number">4.7</span> Review Phyloseq Objects<a class="anchor" aria-label="anchor" href="#review-phyloseq-objects"><i class="fas fa-link"></i></a>
</h2>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://dx.plos.org/10.1371/journal.pone.0061217">phyloseq</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://ape-package.ird.fr/">ape</a></span><span class="op">)</span></span>
<span></span>
<span><span class="va">ps</span> <span class="op">&lt;-</span> <span class="va">ps_dietswap</span></span>
<span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">metagMisc</span><span class="op">)</span></span></code></pre></div>
<div id="taxonomy-rank-names" class="section level3" number="4.7.1">
<h3>
<span class="header-section-number">4.7.1</span> Taxonomy rank names<a class="anchor" aria-label="anchor" href="#taxonomy-rank-names"><i class="fas fa-link"></i></a>
</h3>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/rank_names.html">rank_names</a></span><span class="op">(</span><span class="va">ps</span><span class="op">)</span></span></code></pre></div>
<pre><code>[1] "Phylum" "Family" "Genus" </code></pre>
</div>
<div id="number-of-taxa" class="section level3" number="4.7.2">
<h3>
<span class="header-section-number">4.7.2</span> Number of taxa<a class="anchor" aria-label="anchor" href="#number-of-taxa"><i class="fas fa-link"></i></a>
</h3>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/ntaxa-methods.html">ntaxa</a></span><span class="op">(</span><span class="va">ps</span><span class="op">)</span></span></code></pre></div>
<pre><code>[1] 130</code></pre>
</div>
<div id="split-by-taxon-name-e.g.-firmicutes" class="section level3" number="4.7.3">
<h3>
<span class="header-section-number">4.7.3</span> Split by taxon name, e.g. Firmicutes<a class="anchor" aria-label="anchor" href="#split-by-taxon-name-e.g.-firmicutes"><i class="fas fa-link"></i></a>
</h3>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">taxa_stats</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/metagMisc/man/phyloseq_sep_tax.html">phyloseq_sep_tax</a></span><span class="op">(</span><span class="va">ps</span>, TaxRank <span class="op">=</span> <span class="st">"Phylum"</span>, drop_NA <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/cat.html">cat</a></span><span class="op">(</span><span class="st">"\n"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">taxa_stats</span><span class="op">$</span><span class="va">Firmicutes</span></span></code></pre></div>
<pre><code>phyloseq-class experiment-level object
otu_table()   OTU Table:         [ 76 taxa and 222 samples ]
sample_data() Sample Data:       [ 222 samples by 8 sample variables ]
tax_table()   Taxonomy Table:    [ 76 taxa by 3 taxonomic ranks ]</code></pre>
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/cat.html">cat</a></span><span class="op">(</span><span class="st">"\n"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">taxa_stats</span><span class="op">$</span><span class="va">Bacteroidetes</span></span></code></pre></div>
<pre><code>phyloseq-class experiment-level object
otu_table()   OTU Table:         [ 16 taxa and 222 samples ]
sample_data() Sample Data:       [ 222 samples by 8 sample variables ]
tax_table()   Taxonomy Table:    [ 16 taxa by 3 taxonomic ranks ]</code></pre>
<div class="sourceCode" id="cb51"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/cat.html">cat</a></span><span class="op">(</span><span class="st">"\n"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb52"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">taxa_stats</span><span class="op">$</span><span class="va">Actinobacteria</span></span></code></pre></div>
<pre><code>phyloseq-class experiment-level object
otu_table()   OTU Table:         [ 8 taxa and 222 samples ]
sample_data() Sample Data:       [ 222 samples by 8 sample variables ]
tax_table()   Taxonomy Table:    [ 8 taxa by 3 taxonomic ranks ]</code></pre>
</div>
<div id="number-of-samples" class="section level3" number="4.7.4">
<h3>
<span class="header-section-number">4.7.4</span> Number of samples<a class="anchor" aria-label="anchor" href="#number-of-samples"><i class="fas fa-link"></i></a>
</h3>
<div class="sourceCode" id="cb54"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/nsamples-methods.html">nsamples</a></span><span class="op">(</span><span class="va">ps</span><span class="op">)</span></span></code></pre></div>
<pre><code>[1] 222</code></pre>
</div>
<div id="sample-names" class="section level3" number="4.7.5">
<h3>
<span class="header-section-number">4.7.5</span> Sample names<a class="anchor" aria-label="anchor" href="#sample-names"><i class="fas fa-link"></i></a>
</h3>
<div class="sourceCode" id="cb56"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/sample_names-methods.html">sample_names</a></span><span class="op">(</span><span class="va">ps</span><span class="op">)</span>, <span class="fl">20</span><span class="op">)</span></span></code></pre></div>
<pre><code> [1] "Sample-1"  "Sample-2"  "Sample-3"  "Sample-4"  "Sample-5"  "Sample-6" 
 [7] "Sample-7"  "Sample-8"  "Sample-9"  "Sample-10" "Sample-11" "Sample-12"
[13] "Sample-13" "Sample-14" "Sample-15" "Sample-16" "Sample-17" "Sample-18"
[19] "Sample-19" "Sample-20"</code></pre>
</div>
<div id="sample-variables" class="section level3" number="4.7.6">
<h3>
<span class="header-section-number">4.7.6</span> Sample variables<a class="anchor" aria-label="anchor" href="#sample-variables"><i class="fas fa-link"></i></a>
</h3>
<div class="sourceCode" id="cb58"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/sample_variables.html">sample_variables</a></span><span class="op">(</span><span class="va">ps</span><span class="op">)</span></span></code></pre></div>
<pre><code>[1] "subject"                "sex"                    "nationality"           
[4] "group"                  "sample"                 "timepoint"             
[7] "timepoint.within.group" "bmi_group"             </code></pre>
</div>
<div id="abundance-table" class="section level3" number="4.7.7">
<h3>
<span class="header-section-number">4.7.7</span> Abundance table<a class="anchor" aria-label="anchor" href="#abundance-table"><i class="fas fa-link"></i></a>
</h3>
<div class="sourceCode" id="cb60"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/otu_table-methods.html">otu_table</a></span><span class="op">(</span><span class="va">ps</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span></code></pre></div>
<pre><code>OTU Table:          [5 taxa and 5 samples]
                     taxa are rows
                             Sample-1 Sample-2 Sample-3 Sample-4 Sample-5
Actinomycetaceae                    0        1        0        1        0
Aerococcus                          0        0        0        0        0
Aeromonas                           0        0        0        0        0
Akkermansia                        18       97       67      256       21
Alcaligenes faecalis et rel.        1        2        3        2        2</code></pre>
</div>
<div id="taxonomy-table" class="section level3" number="4.7.8">
<h3>
<span class="header-section-number">4.7.8</span> Taxonomy table<a class="anchor" aria-label="anchor" href="#taxonomy-table"><i class="fas fa-link"></i></a>
</h3>
<div class="sourceCode" id="cb62"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/tax_table-methods.html">tax_table</a></span><span class="op">(</span><span class="va">ps</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="op">]</span></span></code></pre></div>
<pre><code>Taxonomy Table:     [5 taxa by 3 taxonomic ranks]:
                             Phylum            Family           
Actinomycetaceae             "Actinobacteria"  "Actinobacteria" 
Aerococcus                   "Firmicutes"      "Bacilli"        
Aeromonas                    "Proteobacteria"  "Proteobacteria" 
Akkermansia                  "Verrucomicrobia" "Verrucomicrobia"
Alcaligenes faecalis et rel. "Proteobacteria"  "Proteobacteria" 
                             Genus                         
Actinomycetaceae             "Actinomycetaceae"            
Aerococcus                   "Aerococcus"                  
Aeromonas                    "Aeromonas"                   
Akkermansia                  "Akkermansia"                 
Alcaligenes faecalis et rel. "Alcaligenes faecalis et rel."</code></pre>
</div>
<div id="phylogenetic-tree-if-available" class="section level3" number="4.7.9">
<h3>
<span class="header-section-number">4.7.9</span> Phylogenetic tree if available<a class="anchor" aria-label="anchor" href="#phylogenetic-tree-if-available"><i class="fas fa-link"></i></a>
</h3>
</div>
<div id="taxa-names" class="section level3" number="4.7.10">
<h3>
<span class="header-section-number">4.7.10</span> Taxa names<a class="anchor" aria-label="anchor" href="#taxa-names"><i class="fas fa-link"></i></a>
</h3>
<div class="sourceCode" id="cb64"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/taxa_names-methods.html">taxa_names</a></span><span class="op">(</span><span class="va">ps</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span></span></code></pre></div>
<pre><code> [1] "Actinomycetaceae"                  "Aerococcus"                       
 [3] "Aeromonas"                         "Akkermansia"                      
 [5] "Alcaligenes faecalis et rel."      "Allistipes et rel."               
 [7] "Anaerobiospirillum"                "Anaerofustis"                     
 [9] "Anaerostipes caccae et rel."       "Anaerotruncus colihominis et rel."</code></pre>
</div>
<div id="summarize-phyloseq-object" class="section level3" number="4.7.11">
<h3>
<span class="header-section-number">4.7.11</span> Summarize Phyloseq Object<a class="anchor" aria-label="anchor" href="#summarize-phyloseq-object"><i class="fas fa-link"></i></a>
</h3>
<div class="sourceCode" id="cb66"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">microbiome</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/microbiome/man/summarize_phyloseq.html">summarize_phyloseq</a></span><span class="op">(</span><span class="va">ps</span><span class="op">)</span></span></code></pre></div>
<pre><code>[[1]]
[1] "1] Min. number of reads = 1776"

[[2]]
[1] "2] Max. number of reads = 28883"

[[3]]
[1] "3] Total number of reads = 2949085"

[[4]]
[1] "4] Average number of reads = 13284.1666666667"

[[5]]
[1] "5] Median number of reads = 13255"

[[6]]
[1] "7] Sparsity = 0.205647955647956"

[[7]]
[1] "6] Any OTU sum to 1 or less? YES"

[[8]]
[1] "8] Number of singletons = 10"

[[9]]
[1] "9] Percent of OTUs that are singletons \n        (i.e. exactly one read detected across all samples)2.30769230769231"

[[10]]
[1] "10] Number of sample variables are: 8"

[[11]]
[1] "subject"                "sex"                    "nationality"           
[4] "group"                  "sample"                 "timepoint"             
[7] "timepoint.within.group" "bmi_group"             </code></pre>
</div>
<div id="sort-samples-in-ascending-and-descending" class="section level3" number="4.7.12">
<h3>
<span class="header-section-number">4.7.12</span> Sort Samples in ascending and descending<a class="anchor" aria-label="anchor" href="#sort-samples-in-ascending-and-descending"><i class="fas fa-link"></i></a>
</h3>
<div class="sourceCode" id="cb68"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/cat.html">cat</a></span><span class="op">(</span><span class="st">"In ascending order\n"</span><span class="op">)</span></span></code></pre></div>
<pre><code>In ascending order</code></pre>
<div class="sourceCode" id="cb70"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sort.html">sort</a></span><span class="op">(</span><span class="fu">phyloseq</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/sample_sums.html">sample_sums</a></span><span class="op">(</span><span class="va">ps</span><span class="op">)</span>, decreasing <span class="op">=</span> <span class="cn">F</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code> Sample-56 Sample-195 Sample-196  Sample-42 Sample-164 Sample-147 
      1776       4477       4606       5006       5541       5624 </code></pre>
<div class="sourceCode" id="cb72"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/cat.html">cat</a></span><span class="op">(</span><span class="st">"\nIn descending order\n"</span><span class="op">)</span></span></code></pre></div>
<pre><code>
In descending order</code></pre>
<div class="sourceCode" id="cb74"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sort.html">sort</a></span><span class="op">(</span><span class="fu">phyloseq</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/sample_sums.html">sample_sums</a></span><span class="op">(</span><span class="va">ps</span><span class="op">)</span>, decreasing <span class="op">=</span> <span class="cn">T</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code>  Sample-3  Sample-11  Sample-12  Sample-54 Sample-208  Sample-20 
     28883      26895      25238      23529      23131      22413 </code></pre>
</div>
<div id="drop-samples-below-a-threshold" class="section level3" number="4.7.13">
<h3>
<span class="header-section-number">4.7.13</span> Drop Samples Below a Threshold<a class="anchor" aria-label="anchor" href="#drop-samples-below-a-threshold"><i class="fas fa-link"></i></a>
</h3>
<div class="sourceCode" id="cb76"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ps1perc0</span> <span class="op">&lt;-</span> <span class="fu">phyloseq</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/subset_samples-methods.html">subset_samples</a></span><span class="op">(</span><span class="va">ps</span>, <span class="fu">phyloseq</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/sample_sums.html">sample_sums</a></span><span class="op">(</span><span class="va">ps</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">ps1perc0</span></span></code></pre></div>
<pre><code>phyloseq-class experiment-level object
otu_table()   OTU Table:         [ 130 taxa and 222 samples ]
sample_data() Sample Data:       [ 222 samples by 8 sample variables ]
tax_table()   Taxonomy Table:    [ 130 taxa by 3 taxonomic ranks ]</code></pre>
</div>
<div id="drop-taxa-below-a-threshold" class="section level3" number="4.7.14">
<h3>
<span class="header-section-number">4.7.14</span> Drop Taxa Below a Threshold<a class="anchor" aria-label="anchor" href="#drop-taxa-below-a-threshold"><i class="fas fa-link"></i></a>
</h3>
<div class="sourceCode" id="cb78"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pstaxa01</span> <span class="op">&lt;-</span> <span class="fu">phyloseq</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/prune_taxa-methods.html">prune_taxa</a></span><span class="op">(</span><span class="fu">phyloseq</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/taxa_sums.html">taxa_sums</a></span><span class="op">(</span><span class="va">ps</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">1</span>, <span class="va">ps</span><span class="op">)</span></span>
<span><span class="va">pstaxa01</span> </span></code></pre></div>
<pre><code>phyloseq-class experiment-level object
otu_table()   OTU Table:         [ 120 taxa and 222 samples ]
sample_data() Sample Data:       [ 222 samples by 8 sample variables ]
tax_table()   Taxonomy Table:    [ 120 taxa by 3 taxonomic ranks ]</code></pre>

</div>
</div>
</div>



  <div class="chapter-nav">
<div class="prev"><a href="prepare-qiime2-output.html"><span class="header-section-number">3</span> Prepare Qiime2 Output</a></div>
<div class="next"><a href="data-tarnsformation.html"><span class="header-section-number">5</span> Data Tranformation from Phyloseq Objects</a></div>
</div></main><div class="col-md-3 col-lg-2 d-none d-md-block sidebar sidebar-chapter">
    <nav id="toc" data-toggle="toc" aria-label="On this page"><h2>On this page</h2>
      <ul class="nav navbar-nav">
<li><a class="nav-link" href="#import-preprocessed-microbiome-data"><span class="header-section-number">4</span> Import Preprocessed Microbiome Data</a></li>
<li><a class="nav-link" href="#the-globalpatterns-dataset"><span class="header-section-number">4.1</span> The GlobalPatterns dataset</a></li>
<li><a class="nav-link" href="#the-dietswap-dataset"><span class="header-section-number">4.2</span> The dietswap dataset</a></li>
<li><a class="nav-link" href="#caporaso-dataset"><span class="header-section-number">4.3</span> Caporaso dataset</a></li>
<li><a class="nav-link" href="#kostic_crc-dataset"><span class="header-section-number">4.4</span> Kostic_CRC dataset</a></li>
<li><a class="nav-link" href="#save_ps_objects"><span class="header-section-number">4.5</span> Save objects for transformation and exploration</a></li>
<li><a class="nav-link" href="#confirm-saved-objects"><span class="header-section-number">4.6</span> Confirm saved objects</a></li>
<li>
<a class="nav-link" href="#review-phyloseq-objects"><span class="header-section-number">4.7</span> Review Phyloseq Objects</a><ul class="nav navbar-nav">
<li><a class="nav-link" href="#taxonomy-rank-names"><span class="header-section-number">4.7.1</span> Taxonomy rank names</a></li>
<li><a class="nav-link" href="#number-of-taxa"><span class="header-section-number">4.7.2</span> Number of taxa</a></li>
<li><a class="nav-link" href="#split-by-taxon-name-e.g.-firmicutes"><span class="header-section-number">4.7.3</span> Split by taxon name, e.g. Firmicutes</a></li>
<li><a class="nav-link" href="#number-of-samples"><span class="header-section-number">4.7.4</span> Number of samples</a></li>
<li><a class="nav-link" href="#sample-names"><span class="header-section-number">4.7.5</span> Sample names</a></li>
<li><a class="nav-link" href="#sample-variables"><span class="header-section-number">4.7.6</span> Sample variables</a></li>
<li><a class="nav-link" href="#abundance-table"><span class="header-section-number">4.7.7</span> Abundance table</a></li>
<li><a class="nav-link" href="#taxonomy-table"><span class="header-section-number">4.7.8</span> Taxonomy table</a></li>
<li><a class="nav-link" href="#phylogenetic-tree-if-available"><span class="header-section-number">4.7.9</span> Phylogenetic tree if available</a></li>
<li><a class="nav-link" href="#taxa-names"><span class="header-section-number">4.7.10</span> Taxa names</a></li>
<li><a class="nav-link" href="#summarize-phyloseq-object"><span class="header-section-number">4.7.11</span> Summarize Phyloseq Object</a></li>
<li><a class="nav-link" href="#sort-samples-in-ascending-and-descending"><span class="header-section-number">4.7.12</span> Sort Samples in ascending and descending</a></li>
<li><a class="nav-link" href="#drop-samples-below-a-threshold"><span class="header-section-number">4.7.13</span> Drop Samples Below a Threshold</a></li>
<li><a class="nav-link" href="#drop-taxa-below-a-threshold"><span class="header-section-number">4.7.14</span> Drop Taxa Below a Threshold</a></li>
</ul>
</li>
</ul>

      <div class="book-extra">
        <ul class="list-unstyled">
<li><a id="book-source" href="https://github.com/tmbuza/imap-data-preparation.git/blob/gh-pages/04_preprocessed_external_data.Rmd">View source <i class="fas fa-air-freshener"></i></a></li>
          <li><a id="book-edit" href="https://github.com/tmbuza/imap-data-preparation.git/edit/gh-pages/04_preprocessed_external_data.Rmd">Edit this page <i class="fas fa-air-freshener"></i></a></li>
        </ul>
</div>
    </nav>
</div>

</div>
</div> <!-- .container -->

<footer class="bg-primary text-light mt-5"><div class="container">
    <div class="row">
      <div class="col-12 col-md-6 mt-3">
        <p><strong><u><a class="text-light" href="https://microbiome.complexdatainsights.com"><big>IMAP:</big> Integrated Microbiome Analysis Pipelines</a></u></strong>: End-to-End Practical User Guides Using Integrated Approaches</p>
<br><p>Authored by Teresia Mrema-Buza, Last updated: <b>2024-04-13</b>.</p>
<br>
</div>
      <div class="col-12 col-md-6 mt-3">
        <p>Note: This practical guide was created using the <a class="text-light" href="https://bookdown.org/yihui/bookdown/">bookdown</a> R package, styled with the modern bs4_book format.</p>
      </div>

    </div>
  </div>
</footer>
</body>
</html>
